src/libutil/archive.cc

```diff
@@ -171,93 +171,105 @@ static void parse(ParseSink & sink, Source & source, const Path & path)
     s = readString(source);
     if (s != "(") throw badArchive("expected open tag");
 
-    enum { tpUnknown, tpRegular, tpDirectory, tpSymlink } type = tpUnknown;
```

Comment:
The change I did to the parser is basically to replace this extra state variable with control flow. This is needed so that the "content" and "executable" parsing happens within the callback. I then changed the others to keep the code design consistent (and get rid of this enum entirely, as opposed to confusingly just getting rid of `tpRegular`).